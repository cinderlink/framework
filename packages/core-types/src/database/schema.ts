import type Emittery from "emittery";
import type { CID } from "multiformats";
import type { DIDDagInterface } from "../dag";
import type { TableDefinition, TableInterface, TableRow } from "./table";

export type SchemaDef = {
  schemaId: string;
  tables: Record<string, TableDefinition<any>>;
};

export type SchemaEvents = {
  "/schema/loaded": undefined;
};

export type SavedSchema = {
  schemaId: string;
  defs: Record<string, TableDefinition>;
  tables: Record<string, string | undefined>;
};

export interface SchemaInterface extends Emittery<SchemaEvents> {
  tables: Record<string, TableInterface<any>>;
  schemaId: string;
  defs: Record<string, TableDefinition<any>>;
  dag: DIDDagInterface;
  encrypted: boolean;

  createTable<Def extends TableDefinition<any> = TableDefinition<any>>(
    name: string,
    def: Def
  ): Promise<void>;
  dropTable(name: string): Promise<void>;

  getTable<
    Row extends TableRow = TableRow,
    Def extends TableDefinition<Row> = TableDefinition<Row>
  >(
    name: string
  ): TableInterface<Row, Def>;

  setDefs<Def extends TableDefinition = TableDefinition>(
    defs: Record<string, Def>
  ): void;
  save(): Promise<CID | undefined>;
}

export interface SchemaClass {
  new (): SchemaInterface;
  load(
    cid: string | CID,
    dag: DIDDagInterface,
    encrypted: boolean
  ): Promise<SchemaInterface>;
  fromSavedSchema(
    data: SavedSchema,
    dag: DIDDagInterface,
    encrypted: boolean
  ): Promise<SchemaInterface>;
}
