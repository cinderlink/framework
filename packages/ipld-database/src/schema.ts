import { DIDDagInterface } from "@candor/core-types";
import Emittery from "emittery";
import type { CID } from "multiformats";
import { Table, TableDefinition, TableRow } from "./table";

export type SavedSchema = {
  name: string;
  defs: Record<string, TableDefinition>;
  tables: Record<string, string | undefined>;
};

export type SchemaEvents = {
  "/schema/loaded": undefined;
};

export class Schema extends Emittery<SchemaEvents> {
  public tables: Record<string, Table> = {};
  constructor(
    public name: string,
    public defs: Record<string, TableDefinition>,
    public dag: DIDDagInterface,
    public encrypted = true
  ) {
    super();
    Object.entries(defs).forEach(([name, def]) => {
      this.tables[name] = new Table(def, this.dag);
    });
  }

  async createTable(name: string, def: TableDefinition) {
    this.tables[name] = new Table(def, this.dag);
  }

  async dropTable(name: string) {
    delete this.tables[name];
  }

  getTable<T extends TableRow = TableRow>(name: string) {
    return this.tables[name] as Table<T>;
  }

  async save() {
    console.info("saving schema", this.name);
    const tables: Record<string, string | undefined> = {};
    await Promise.all(
      Object.entries(this.tables).map(async ([name]) => {
        const tableCID = await this.tables[name].save();
        if (tableCID) {
          tables[name] = tableCID?.toString();
        }
      })
    );
    console.info("storing schema", this.name, this.defs, tables);
    const savedSchema = {
      name: this.name,
      defs: this.defs,
      tables,
    };
    return this.encrypted
      ? this.dag.storeEncrypted(savedSchema)
      : this.dag.store(savedSchema);
  }

  static async load(cid: string | CID, dag: DIDDagInterface, encrypted = true) {
    const data = encrypted
      ? await dag.loadDecrypted<SavedSchema>(cid)
      : await dag.load<SavedSchema>(cid);
    return Schema.fromSavedSchema(data, dag);
  }

  static async fromSavedSchema(
    data: SavedSchema,
    dag: DIDDagInterface,
    encrypted = true
  ) {
    if (!data) throw new Error("Invalid schema data");
    console.info("loading from saved schema", data);
    const schema = new Schema(data.name, data.defs, dag, encrypted);
    await Promise.all(
      Object.entries(data.tables).map(async ([name, tableCID]) => {
        console.info("loading table", { name, tableCID });
        await schema.tables[name]?.load(tableCID!);
      })
    );
    schema.emit("/schema/loaded");
    return schema;
  }
}
