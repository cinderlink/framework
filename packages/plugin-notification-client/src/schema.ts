import { CandorClientInterface, TableDefinition } from "@candor/core-types";
import { Schema } from "@candor/ipld-database";

export const NotificationSchemaDef: Record<string, TableDefinition> = {
  notifications: {
    schemaId: "notification",
    encrypted: false,
    aggregate: {},
    indexes: {
      id: {
        unique: true,
        fields: ["id"],
      },
    },
    rollup: 1000,
    searchOptions: {
      fields: ["id", "type"],
    },
    schema: {
      type: "object",
      properties: {
        id: { type: "string" },
        type: { type: "string" },
        title: { type: "string" },
        body: { type: "string" },
        dismissed: { type: "boolean" },
        createdAt: { type: "number" },
        read: { type: "boolean" },
        link: { type: "string" },
        metaData: { type: "object" },
      },
    },
  },
};

export async function loadNotificationSchema(
  client: CandorClientInterface<any>
) {
  console.log(`plugin/notification/client > preparing schema`);
  if (!client.schemas["notification"]) {
    const schema = new Schema(
      "notification",
      NotificationSchemaDef,
      client.dag
    );
    await client.addSchema("notification", schema);
  } else {
    client.schemas["notification"].setDefs(NotificationSchemaDef);
  }
}

export default NotificationSchemaDef;
