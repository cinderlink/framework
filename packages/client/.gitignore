candor/
.turbo/
dist/
node_modules/