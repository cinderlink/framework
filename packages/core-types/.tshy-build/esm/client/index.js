export * from "./types";
export * from "./interface";
//# sourceMappingURL=index.js.map