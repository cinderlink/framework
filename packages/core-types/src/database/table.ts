import type { SchemaObject } from "ajv";
import type { CID } from "multiformats";
import type Emittery from "emittery";
import type {
  default as Minisearch,
  AsPlainObject,
  Options as SearchOptions,
} from "minisearch";
import type { DIDDagInterface } from "../dag";

export type TableRow<Data extends Record<string, unknown> = {}> = {
  id?: number;
  [key: string]: unknown;
} & Data;

export type TableDefinition = {
  encrypted: boolean;
  schema: SchemaObject;
  indexes: string[];
  aggregate: {
    [key: string]:
      | "sum"
      | "count"
      | "avg"
      | "min"
      | "max"
      | "distinct"
      | "range";
  };
  searchOptions: SearchOptions;
  rollup: number;
};

export type TableBlock<T extends TableRow = TableRow> = {
  prevCID?: string;
  fromIndex: number;
  toIndex: number;
  records: Record<number, T>;
  indexes: Record<string, Record<string, number[]>>;
  aggregates: Record<string, number | string | string[] | [number, number]>;
  search?: Minisearch;
  index?: AsPlainObject;
};

export type TableEvents = {
  "/table/loaded": undefined;
};

export interface TableInterface<T extends TableRow = TableRow>
  extends Emittery<TableEvents> {
  currentIndex: number;
  currentBlock: TableBlock;
  encrypted: boolean;
  def: TableDefinition;
  dag: DIDDagInterface;

  createBlock(prevCID: string | undefined): TableBlock;
  insert(data: T): Promise<number>;
  search(query: string, limit: number): Promise<T[]>;
  rollup(): Promise<void>;
  updateBlockAggregates(block: TableBlock): void;
  save(): Promise<string | undefined>;
  update(id: number, data: Partial<T>): Promise<number>;
  load(cid: CID | string): Promise<void>;
  unwind(
    test: (block: TableBlock, cid: string | undefined) => Promise<boolean>
  ): Promise<void>;
  findByIndex(index: string, value: string): Promise<T | undefined>;
  where(match: (row: T) => boolean): Promise<T[]>;
  find(match: (row: T) => boolean): Promise<T | undefined>;
  upsert(index: string, value: string | number, data: T): Promise<number>;
  assertValid(data: T): void;
  isValid(data: T): boolean;
}
