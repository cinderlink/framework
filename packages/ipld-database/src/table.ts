import type { DIDDagInterface } from "@candor/core-types";
import { CID } from "multiformats";
import Emittery from "emittery";
import Ajv from "ajv";
import Minisearch from "minisearch";
import {
  TableBlock,
  TableDefinition,
  TableEvents,
  TableInterface,
  TableRow,
} from "@candor/core-types/src/database/table";

const ajv = new Ajv();

export class Table<T extends TableRow = TableRow>
  extends Emittery<TableEvents>
  implements TableInterface<T>
{
  currentIndex: number = 0;
  currentBlock: TableBlock;
  encrypted: boolean;

  constructor(public def: TableDefinition, public dag: DIDDagInterface) {
    super();
    this.encrypted = def.encrypted;
    this.currentBlock = this.createBlock();
  }

  createBlock(prevCID: string | undefined = this.currentBlock?.prevCID) {
    return {
      prevCID,
      fromIndex: this.currentIndex,
      toIndex: this.currentIndex + this.def.rollup,
      records: {},
      indexes: {},
      aggregates: {},
      search: new Minisearch(this.def.searchOptions),
    };
  }

  async insert(data: T) {
    this.assertValid(data);
    const rowCID = this.encrypted
      ? await this.dag.storeEncrypted(data)
      : await this.dag.store(data);
    if (!rowCID) {
      throw new Error("Failed to store row");
    }
    data.id = this.currentIndex++;
    this.currentBlock.records[data.id] = data;
    this.def.indexes.forEach((index) => {
      if (!this.currentBlock.indexes[index]) {
        this.currentBlock.indexes[index] = {};
      }
      const value = data[index] as string;
      if (!this.currentBlock.indexes[index][value]) {
        this.currentBlock.indexes[index][value] = [data.id as number];
      } else {
        this.currentBlock.indexes[index][value].push(data.id as number);
      }
    });
    this.currentBlock.search?.add(data);
    if (this.currentIndex - this.currentBlock.fromIndex >= this.def.rollup) {
      await this.rollup();
    }

    return data.id;
  }

  async search(query: string, limit = 10) {
    let results: T[] = [];
    await this.unwind(async (block) => {
      const searchResults = block.search?.search(query);
      const searchIds = [...new Set(searchResults?.map((r) => r.id) || [])];
      const records: T[] = searchIds.map((id) => block.records[id] as T);
      if (records) {
        results = results.concat(records);
      }
      return results.length < limit;
    });
    return results.slice(0, limit);
  }

  async rollup() {
    this.currentBlock = this.updateBlockAggregates(this.currentBlock);
    this.currentBlock.index = this.currentBlock.search?.toJSON();
    this.currentBlock.toIndex = this.currentIndex;
    delete this.currentBlock.search;
    if (!this.currentBlock.prevCID) {
      // can't save undefined
      delete this.currentBlock.prevCID;
    }
    const blockCID = this.encrypted
      ? await this.dag.storeEncrypted(this.currentBlock)
      : await this.dag.store(this.currentBlock);
    this.currentBlock = this.createBlock(blockCID?.toString());
  }

  updateBlockAggregates(block: TableBlock) {
    block.aggregates = {};
    const records = Object.values(block.records);
    Object.entries(this.def.aggregate).forEach(([key, type]) => {
      if (type === "sum") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] as number),
          0
        );
      } else if (type === "count") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] ? 1 : 0),
          0
        );
      } else if (type === "avg") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] as number),
          0
        );
      } else if (type === "min") {
        block.aggregates[key] = Math.min(
          ...records.map((row) => row[key] as number)
        );
      } else if (type === "max") {
        block.aggregates[key] = Math.max(
          ...records.map((row) => row[key] as number)
        );
      } else if (type === "distinct") {
        block.aggregates[key] = [
          ...new Set(records.map((row) => row[key] as string)),
        ];
      } else if (type === "range") {
        block.aggregates[key] = [
          Math.min(...records.map((row) => row[key] as number)),
          Math.max(...records.map((row) => row[key] as number)),
        ];
      }
    });
    return block;
  }

  async save() {
    if (Object.values(this.currentBlock.records).length > 0) {
      await this.rollup();
    }
    return this.currentBlock.prevCID;
  }

  async update(id: number, data: Partial<T>) {
    // we want to seek backwards from the current block
    // to find the block containing the row with the given id
    // we then want to update the row in that block
    // and then re-write that block and all subsequent blocks
    // to update the indexes, aggregates, and CID references
    // we also want to update the search index

    let blocksUnwound: string[] = [];
    let rewriteBlock: TableBlock | undefined = undefined;

    const unwindFn = async (block: TableBlock, cid: string | undefined) => {
      // if the block contains the row we want to update
      const row = block.records[id];
      const updated = { ...row, ...data };
      if (row) {
        block.records[id] = updated;
        // update the indexes
        this.def.indexes.forEach((index) => {
          // if the index has changed
          if (updated[index] && updated[index] !== row[index]) {
            // remove the row from the old index
            block.indexes[index][row[index] as string] = block.indexes[index][
              row[index] as string
            ].filter((rowID) => rowID !== id);
            // add the row to the new index
            if (!block.indexes[index][updated[index] as string]) {
              block.indexes[index][updated[index] as string] = [id];
            } else {
              block.indexes[index][updated[index] as string].push(id);
            }
          }
        });
        if (cid) {
          // update the aggregates
          block = this.updateBlockAggregates(block);
          // update the search index
          block.search?.removeAll();
          block.search?.addAll(Object.values(block.records));
          // update the CID reference
          rewriteBlock = block;
        }
        return false;
      }
      return true;
    };

    await this.unwind(unwindFn);

    // rewrite the blocks
    if (rewriteBlock) {
      let prevCID = this.encrypted
        ? await this.dag.storeEncrypted(rewriteBlock)
        : await this.dag.store(rewriteBlock);
      for (const cid of blocksUnwound) {
        const block = this.encrypted
          ? await this.dag.loadDecrypted<TableBlock>(cid)
          : await this.dag.load<TableBlock>(cid);
        if (block) {
          block.prevCID = prevCID?.toString();
          prevCID = this.encrypted
            ? await this.dag.storeEncrypted(block)
            : await this.dag.store(block);
        }
      }
    }

    return id;
  }

  async load(cid: CID | string) {
    const block = this.encrypted
      ? await this.dag.loadDecrypted<TableBlock>(cid)
      : await this.dag.load<TableBlock>(cid);
    if (block) {
      this.currentBlock = block;
      this.currentBlock.search = this.currentBlock.index
        ? await Minisearch.loadJS(
            this.currentBlock.index,
            this.def.searchOptions
          )
        : new Minisearch(this.def.searchOptions);
      this.currentIndex = block.toIndex;
      // this.currentBlock = this.createBlock(cid.toString());
    } else {
      this.currentBlock = this.createBlock();
    }
    this.emit("/table/loaded");
  }

  async unwind(
    test: (block: TableBlock, cid: string | undefined) => Promise<boolean>
  ) {
    let cid: string | undefined = undefined;
    let block: TableBlock | undefined = this.currentBlock;
    let prevented = false;
    while (prevented === false && block) {
      prevented = !(await test(block, cid).catch((err) => {
        console.error("unwind error", err);
        return false;
      }));
      if (prevented) {
        cid = undefined;
        block = undefined;
        return;
      }
      cid = block.prevCID;
      block = block.prevCID
        ? this.encrypted
          ? await this.dag.loadDecrypted<TableBlock>(block.prevCID)
          : await this.dag.load<TableBlock>(block.prevCID)
        : undefined;
    }
  }

  async findByIndex(index: string, value: string) {
    let result: T | undefined;
    await this.unwind(async (block) => {
      if (block.indexes[index]?.[value]) {
        for (let rowID of block.indexes[index][value]) {
          result = block.records[rowID] as T;
          if (result) {
            return false;
          }
        }
      }
      return true;
    });
    return result;
  }

  async where(match: (row: T) => boolean) {
    const results: T[] = [];
    await this.unwind(async (block) => {
      for (let row of Object.values(block.records)) {
        if (match(row as T)) {
          results.push(row as T);
        }
      }
      return true;
    });
    return results;
  }

  async find(match: (row: T) => boolean) {
    let result: T | undefined;
    await this.unwind(async (block) => {
      for (let row of Object.values(block.records)) {
        if (match(row as T)) {
          result = row as T;
          return false;
        }
      }
      return true;
    });
    return result;
  }

  async upsert(index: string, value: string | number, data: T) {
    const existing = await this.findByIndex(index, value as string);
    if (existing?.id) {
      return this.update(existing.id, data);
    }
    const record = { [index]: value, ...data };
    return this.insert(record);
  }

  assertValid(data: T) {
    const validate = ajv.compile(this.def.schema);
    const valid = validate(data);
    if (!valid) {
      throw new Error(
        `Invalid data: ${validate.errors?.map(
          (e) => `(${e.instancePath}: ${e.message})`
        )}`
      );
    }
  }

  isValid(data: T) {
    const validate = ajv.compile(this.def.schema);
    return validate(data);
  }
}
