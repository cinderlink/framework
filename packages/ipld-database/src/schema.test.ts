import { TestDIDDag } from "@cryptids/dag-interface";
import { number, object, string } from "superstruct";
import type { DID } from "dids";
import { beforeAll, beforeEach, describe, expect, it } from "vitest";
import { Schema } from "./schema";
import { TableDefinition } from "./table";
import { createCryptidsSeed, createDID } from "@cryptids/client";

const tableDefinition: TableDefinition = {
  indexes: ["name", "id"],
  aggregate: {
    count: "max",
  },
  schema: object({
    name: string(),
    count: number(),
  }),
  searchOptions: {
    fields: ["name", "id"],
  },
  rollup: 10,
};

let did: DID;
let dag: TestDIDDag;

describe("@cryptids/ipld-database/schema", () => {
  beforeAll(async () => {
    const seed = await createCryptidsSeed("test seed");
    did = await createDID(seed);
  });

  beforeEach(async () => {
    dag = new TestDIDDag(did);
  });

  it("should create tables", async () => {
    const schema = new Schema("test", { test: tableDefinition }, dag);
    expect(schema.tables.test?.currentBlock).toMatchInlineSnapshot(`
      {
        "aggregates": {},
        "indexes": {},
        "prevCID": undefined,
        "rows": [],
        "search": {
          "averageFieldLength": [],
          "dirtCount": 0,
          "documentCount": 0,
          "documentIds": {},
          "fieldIds": {
            "id": 1,
            "name": 0,
          },
          "fieldLength": {},
          "index": [],
          "nextId": 0,
          "serializationVersion": 2,
          "storedFields": {},
        },
      }
    `);
  });

  it("should save and restore from dag", async () => {
    const schema = new Schema("test", { test: tableDefinition }, dag);
    for (let i = 0; i < 15; i++) {
      await schema.tables.test?.insert({ name: `test #${i}`, count: i });
    }
    const cid = await schema.save();
    expect(cid).not.toBeUndefined();
    if (!cid) throw new Error("CID is undefined");
    const restored = await Schema.load(cid, dag);
    expect(restored.tables.test?.currentBlock.aggregates).toMatchObject(
      schema.tables.test?.currentBlock.aggregates
    );
    expect(restored.tables.test?.currentBlock.indexes).toMatchObject(
      schema.tables.test?.currentBlock.indexes
    );
    expect(restored.tables.test?.currentBlock.rows).toMatchObject(
      schema.tables.test?.currentBlock.rows
    );
  });
});
