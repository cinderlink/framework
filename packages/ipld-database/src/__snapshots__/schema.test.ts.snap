// Vitest Snapshot v1

exports[`@candor/ipld-database/schema > should create tables 1`] = `
{
  "filters": {
    "search": {
      "averageFieldLength": [],
      "dirtCount": 0,
      "documentCount": 0,
      "documentIds": {},
      "fieldIds": {
        "id": 1,
        "name": 0,
      },
      "fieldLength": {},
      "index": [],
      "nextId": 0,
      "serializationVersion": 2,
      "storedFields": {},
    },
  },
  "headers": {
    "encrypted": false,
    "index": 0,
    "recordsFrom": 1,
    "recordsTo": 1,
    "schema": "test",
    "table": "test",
  },
  "prevCID": undefined,
}
`;

exports[`@candor/ipld-database/schema > should save and restore from dag 1`] = `
Table {
  "currentBlock": {
    "filters": {
      "aggregates": {
        "count": 14,
      },
      "indexes": {
        "name": {
          "z2ikNK1j9q8xKhdHit": {
            "ids": [
              11,
            ],
            "values": [
              "test #10",
            ],
          },
          "z2ikNK1j9q8xKhddCp": {
            "ids": [
              12,
            ],
            "values": [
              "test #11",
            ],
          },
          "z2ikNK1j9q8xKhdxgk": {
            "ids": [
              13,
            ],
            "values": [
              "test #12",
            ],
          },
          "z2ikNK1j9q8xKheJAg": {
            "ids": [
              14,
            ],
            "values": [
              "test #13",
            ],
          },
          "z2ikNK1j9q8xKhedec": {
            "ids": [
              15,
            ],
            "values": [
              "test #14",
            ],
          },
        },
      },
      "search": {
        "averageFieldLength": [
          2,
          1,
        ],
        "dirtCount": 0,
        "documentCount": 5,
        "documentIds": {
          "0": 11,
          "1": 12,
          "2": 13,
          "3": 14,
          "4": 15,
        },
        "fieldIds": {
          "id": 1,
          "name": 0,
        },
        "fieldLength": {
          "0": [
            2,
            1,
          ],
          "1": [
            2,
            1,
          ],
          "2": [
            2,
            1,
          ],
          "3": [
            2,
            1,
          ],
          "4": [
            2,
            1,
          ],
        },
        "index": [
          [
            "15",
            {
              "1": {
                "4": 1,
              },
            },
          ],
          [
            "14",
            {
              "0": {
                "4": 1,
              },
              "1": {
                "3": 1,
              },
            },
          ],
          [
            "13",
            {
              "0": {
                "3": 1,
              },
              "1": {
                "2": 1,
              },
            },
          ],
          [
            "12",
            {
              "0": {
                "2": 1,
              },
              "1": {
                "1": 1,
              },
            },
          ],
          [
            "11",
            {
              "0": {
                "1": 1,
              },
              "1": {
                "0": 1,
              },
            },
          ],
          [
            "10",
            {
              "0": {
                "0": 1,
              },
            },
          ],
          [
            "test",
            {
              "0": {
                "0": 1,
                "1": 1,
                "2": 1,
                "3": 1,
                "4": 1,
              },
            },
          ],
        ],
        "nextId": 5,
        "serializationVersion": 2,
        "storedFields": {},
      },
    },
    "headers": {
      "encrypted": false,
      "index": 0,
      "recordsFrom": 11,
      "recordsTo": 15,
      "schema": "test",
      "table": "test",
    },
    "prevCID": "bagaaieralnu77mm3dv7pzhqdx75kkhrxm4rdlzqb525xa3mndlnrrewl2x5a",
    "records": {
      "11": {
        "count": 10,
        "id": 11,
        "name": "test #10",
      },
      "12": {
        "count": 11,
        "id": 12,
        "name": "test #11",
      },
      "13": {
        "count": 12,
        "id": 13,
        "name": "test #12",
      },
      "14": {
        "count": 13,
        "id": 14,
        "name": "test #13",
      },
      "15": {
        "count": 14,
        "id": 15,
        "name": "test #14",
      },
    },
  },
  "currentBlockIndex": 0,
  "currentIndex": 16,
  "dag": TestDIDDag {
    "dag": TestDag {
      "cache": {
        "bagaaiera4cuuwkkqaouspd62bczp43yozyf3esnpaqymev3wnnpq5hxzwd3q": {
          "defs": {
            "test": {
              "aggregate": {
                "count": "max",
              },
              "encrypted": false,
              "indexes": {
                "name": {
                  "fields": [
                    "name",
                  ],
                  "unique": true,
                },
              },
              "rollup": 10,
              "schema": {
                "properties": {
                  "count": {
                    "type": "number",
                  },
                  "name": {
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "schemaId": "test",
              "searchOptions": {
                "fields": [
                  "name",
                  "id",
                ],
              },
            },
          },
          "schemaId": "test",
          "tables": {
            "test": "bagaaierapykbw62b6mpsfxtmupkefa2uklrxkg4mzb6kcpvyeheejwasz67a",
          },
        },
        "bagaaieralnu77mm3dv7pzhqdx75kkhrxm4rdlzqb525xa3mndlnrrewl2x5a": {
          "filters": {
            "aggregates": {
              "count": 9,
            },
            "indexes": {
              "name": {
                "zPbkLwYJQbRmZQwe": {
                  "ids": [
                    1,
                  ],
                  "values": [
                    "test #0",
                  ],
                },
                "zPbkLwYJQbRmZkRa": {
                  "ids": [
                    2,
                  ],
                  "values": [
                    "test #1",
                  ],
                },
                "zPbkLwYJQbRma5uW": {
                  "ids": [
                    3,
                  ],
                  "values": [
                    "test #2",
                  ],
                },
                "zPbkLwYJQbRmaRPS": {
                  "ids": [
                    4,
                  ],
                  "values": [
                    "test #3",
                  ],
                },
                "zPbkLwYJQbRmaksN": {
                  "ids": [
                    5,
                  ],
                  "values": [
                    "test #4",
                  ],
                },
                "zPbkLwYJQbRmb6MJ": {
                  "ids": [
                    6,
                  ],
                  "values": [
                    "test #5",
                  ],
                },
                "zPbkLwYJQbRmbRqE": {
                  "ids": [
                    7,
                  ],
                  "values": [
                    "test #6",
                  ],
                },
                "zPbkLwYJQbRmbmKA": {
                  "ids": [
                    8,
                  ],
                  "values": [
                    "test #7",
                  ],
                },
                "zPbkLwYJQbRmc6o6": {
                  "ids": [
                    9,
                  ],
                  "values": [
                    "test #8",
                  ],
                },
                "zPbkLwYJQbRmcSH2": {
                  "ids": [
                    10,
                  ],
                  "values": [
                    "test #9",
                  ],
                },
              },
            },
            "search": {
              "averageFieldLength": [
                2,
                1,
              ],
              "dirtCount": 0,
              "documentCount": 10,
              "documentIds": {
                "0": 1,
                "1": 2,
                "2": 3,
                "3": 4,
                "4": 5,
                "5": 6,
                "6": 7,
                "7": 8,
                "8": 9,
                "9": 10,
              },
              "fieldIds": {
                "id": 1,
                "name": 0,
              },
              "fieldLength": {
                "0": [
                  2,
                  1,
                ],
                "1": [
                  2,
                  1,
                ],
                "2": [
                  2,
                  1,
                ],
                "3": [
                  2,
                  1,
                ],
                "4": [
                  2,
                  1,
                ],
                "5": [
                  2,
                  1,
                ],
                "6": [
                  2,
                  1,
                ],
                "7": [
                  2,
                  1,
                ],
                "8": [
                  2,
                  1,
                ],
                "9": [
                  2,
                  1,
                ],
              },
              "index": [
                [
                  "9",
                  {
                    "0": {
                      "9": 1,
                    },
                    "1": {
                      "8": 1,
                    },
                  },
                ],
                [
                  "8",
                  {
                    "0": {
                      "8": 1,
                    },
                    "1": {
                      "7": 1,
                    },
                  },
                ],
                [
                  "7",
                  {
                    "0": {
                      "7": 1,
                    },
                    "1": {
                      "6": 1,
                    },
                  },
                ],
                [
                  "6",
                  {
                    "0": {
                      "6": 1,
                    },
                    "1": {
                      "5": 1,
                    },
                  },
                ],
                [
                  "5",
                  {
                    "0": {
                      "5": 1,
                    },
                    "1": {
                      "4": 1,
                    },
                  },
                ],
                [
                  "4",
                  {
                    "0": {
                      "4": 1,
                    },
                    "1": {
                      "3": 1,
                    },
                  },
                ],
                [
                  "3",
                  {
                    "0": {
                      "3": 1,
                    },
                    "1": {
                      "2": 1,
                    },
                  },
                ],
                [
                  "2",
                  {
                    "0": {
                      "2": 1,
                    },
                    "1": {
                      "1": 1,
                    },
                  },
                ],
                [
                  "10",
                  {
                    "1": {
                      "9": 1,
                    },
                  },
                ],
                [
                  "1",
                  {
                    "0": {
                      "1": 1,
                    },
                    "1": {
                      "0": 1,
                    },
                  },
                ],
                [
                  "0",
                  {
                    "0": {
                      "0": 1,
                    },
                  },
                ],
                [
                  "test",
                  {
                    "0": {
                      "0": 1,
                      "1": 1,
                      "2": 1,
                      "3": 1,
                      "4": 1,
                      "5": 1,
                      "6": 1,
                      "7": 1,
                      "8": 1,
                      "9": 1,
                    },
                  },
                ],
              ],
              "nextId": 10,
              "serializationVersion": 2,
              "storedFields": {},
            },
          },
          "headers": {
            "encrypted": false,
            "index": 0,
            "recordsFrom": 1,
            "recordsTo": 10,
            "schema": "test",
            "table": "test",
          },
          "prevCID": undefined,
          "records": {
            "1": {
              "count": 0,
              "id": 1,
              "name": "test #0",
            },
            "10": {
              "count": 9,
              "id": 10,
              "name": "test #9",
            },
            "2": {
              "count": 1,
              "id": 2,
              "name": "test #1",
            },
            "3": {
              "count": 2,
              "id": 3,
              "name": "test #2",
            },
            "4": {
              "count": 3,
              "id": 4,
              "name": "test #3",
            },
            "5": {
              "count": 4,
              "id": 5,
              "name": "test #4",
            },
            "6": {
              "count": 5,
              "id": 6,
              "name": "test #5",
            },
            "7": {
              "count": 6,
              "id": 7,
              "name": "test #6",
            },
            "8": {
              "count": 7,
              "id": 8,
              "name": "test #7",
            },
            "9": {
              "count": 8,
              "id": 9,
              "name": "test #8",
            },
          },
        },
        "bagaaierapykbw62b6mpsfxtmupkefa2uklrxkg4mzb6kcpvyeheejwasz67a": {
          "filters": {
            "aggregates": {
              "count": 14,
            },
            "indexes": {
              "name": {
                "z2ikNK1j9q8xKhdHit": {
                  "ids": [
                    11,
                  ],
                  "values": [
                    "test #10",
                  ],
                },
                "z2ikNK1j9q8xKhddCp": {
                  "ids": [
                    12,
                  ],
                  "values": [
                    "test #11",
                  ],
                },
                "z2ikNK1j9q8xKhdxgk": {
                  "ids": [
                    13,
                  ],
                  "values": [
                    "test #12",
                  ],
                },
                "z2ikNK1j9q8xKheJAg": {
                  "ids": [
                    14,
                  ],
                  "values": [
                    "test #13",
                  ],
                },
                "z2ikNK1j9q8xKhedec": {
                  "ids": [
                    15,
                  ],
                  "values": [
                    "test #14",
                  ],
                },
              },
            },
            "search": {
              "averageFieldLength": [
                2,
                1,
              ],
              "dirtCount": 0,
              "documentCount": 5,
              "documentIds": {
                "0": 11,
                "1": 12,
                "2": 13,
                "3": 14,
                "4": 15,
              },
              "fieldIds": {
                "id": 1,
                "name": 0,
              },
              "fieldLength": {
                "0": [
                  2,
                  1,
                ],
                "1": [
                  2,
                  1,
                ],
                "2": [
                  2,
                  1,
                ],
                "3": [
                  2,
                  1,
                ],
                "4": [
                  2,
                  1,
                ],
              },
              "index": [
                [
                  "15",
                  {
                    "1": {
                      "4": 1,
                    },
                  },
                ],
                [
                  "14",
                  {
                    "0": {
                      "4": 1,
                    },
                    "1": {
                      "3": 1,
                    },
                  },
                ],
                [
                  "13",
                  {
                    "0": {
                      "3": 1,
                    },
                    "1": {
                      "2": 1,
                    },
                  },
                ],
                [
                  "12",
                  {
                    "0": {
                      "2": 1,
                    },
                    "1": {
                      "1": 1,
                    },
                  },
                ],
                [
                  "11",
                  {
                    "0": {
                      "1": 1,
                    },
                    "1": {
                      "0": 1,
                    },
                  },
                ],
                [
                  "10",
                  {
                    "0": {
                      "0": 1,
                    },
                  },
                ],
                [
                  "test",
                  {
                    "0": {
                      "0": 1,
                      "1": 1,
                      "2": 1,
                      "3": 1,
                      "4": 1,
                    },
                  },
                ],
              ],
              "nextId": 5,
              "serializationVersion": 2,
              "storedFields": {},
            },
          },
          "headers": {
            "encrypted": false,
            "index": 0,
            "recordsFrom": 11,
            "recordsTo": 15,
            "schema": "test",
            "table": "test",
          },
          "prevCID": "bagaaieralnu77mm3dv7pzhqdx75kkhrxm4rdlzqb525xa3mndlnrrewl2x5a",
          "records": {
            "11": {
              "count": 10,
              "id": 11,
              "name": "test #10",
            },
            "12": {
              "count": 11,
              "id": 12,
              "name": "test #11",
            },
            "13": {
              "count": 12,
              "id": 13,
              "name": "test #12",
            },
            "14": {
              "count": 13,
              "id": 14,
              "name": "test #13",
            },
            "15": {
              "count": 14,
              "id": 15,
              "name": "test #14",
            },
          },
        },
      },
    },
    "did": DID {
      "_client": RPCClient {},
      "_id": "did:key:z6MkofP3UGcgQT1iXWFmxczLNCddG5MrWbsrKRNGcJSVyebo",
      "_resolver": Resolver {
        "cache": [Function],
        "registry": {
          "key": [Function],
        },
      },
    },
  },
  "debug": {
    "enabled": false,
    "logger": [Function],
  },
  "def": {
    "aggregate": {
      "count": "max",
    },
    "encrypted": false,
    "indexes": {
      "name": {
        "fields": [
          "name",
        ],
        "unique": true,
      },
    },
    "rollup": 10,
    "schema": {
      "properties": {
        "count": {
          "type": "number",
        },
        "name": {
          "type": "string",
        },
      },
      "type": "object",
    },
    "schemaId": "test",
    "searchOptions": {
      "fields": [
        "name",
        "id",
      ],
    },
  },
  "encrypted": false,
  "tableId": "test",
  "writeStartAt": 0,
  "writing": false,
}
`;
